scripts/test-react.js

```diff
@@ -0,0 +1,179 @@
+}
+
+async function runTest(directory, name) {
+  let source = fs.readFileSync(path.join(reactTestRoot, directory, name)).toString();
+  let { compiledSource } = compileSourceWithPrepack(source);
+
+  let A = runSource(source);
+  expect(typeof A).toBe("function");
+  let B = runSource(compiledSource);
+  expect(typeof B).toBe("function");
+
+  let rendererA = ReactTestRenderer.create(null);
+  let rendererB = ReactTestRenderer.create(null);
+
+  if (A == null || B == null) {
+    throw new Error("React test runner issue");
+  }
+  // // Use the original version of the test in case transforming messes it up.
+  let { getTrials } = A;
+  // // Run tests that assert the rendered output matches.
+  let resultA = getTrials(rendererA, A);
+  let resultB = getTrials(rendererB, B);
+
+  // // the test has returned many values for us to check
+  if (Array.isArray(resultA) && Array.isArray(resultA[0])) {
+    for (let i = 0; i < resultA.length; i++) {
+      let [nameA, valueA] = resultA[i];
+      let [nameB, valueB] = resultB[i];
+      expect(normalize(valueB)).toEqual(normalize(valueA));
+      expect(nameB).toEqual(nameA);
+    }
+  } else {
+    let [nameA, valueA] = resultA;
+    let [nameB, valueB] = resultB;
+    expect(normalize(valueB)).toEqual(normalize(valueA));
+    expect(nameB).toEqual(nameA);
+  }
+}
+
+// Jest tests
+let originalConsoleError = console.error;
+
+describe("Test React", () => {
+  describe("Functional component folding", () => {
+    let directory = "functional-components";
+
+    it("Simple", async () => {
+      await runTest(directory, "simple.js");
+    });
+
+    it("Simple children", async () => {
+      await runTest(directory, "simple-children.js");
+    });
+
+    it("Simple refs", async () => {
+      await runTest(directory, "simple-refs.js");
+    });
+
+    it("Conditional", async () => {
+      await runTest(directory, "conditional.js");
+    });
+
+    it("Key nesting", async () => {
+      await runTest(directory, "key-nesting.js");
+    });
+
+    it("Key change", async () => {
+      await runTest(directory, "key-change.js");
+    });
+
+    it("Component type change", async () => {
+      await runTest(directory, "type-change.js");
+    });
+
+    it("Dynamic props", async () => {
+      await runTest(directory, "dynamic-props.js");
+    });
+
+    it("Dynamic context", async () => {
+      await runTest(directory, "dynamic-context.js");
+    });
+
+    it("React.cloneElement", async () => {
+      await runTest(directory, "clone-element.js");
+    });
+
+    it("Return text", async () => {
+      await runTest(directory, "return-text.js");
+    });
+
+    it("Return undefined", async () => {
+      // this test will cause a React console.error to show
+      // we monkey patch it to stop it polluting the test output
+      // with a false-negative error
+      global.console.error = () => {};
+      await runTest(directory, "return-undefined.js");
+      global.console.error = originalConsoleError;
```

Comment:
Maybe `try { } finally { }` so this is isolated from other tests in case it fails?

Reply:
I thought this wasn't actual JS error that is occurring, rather it's `console.error` being fired?

Reply:
If `runTest` fails for some reason, `global.console.error` won't be restored.

If you do

```js
global.console.error = ...
try {
  doStuff
} finally {
  global.console.error = original
}
```

it is restored no matter what.